Sprint 6

AntonyukV516 opened the pull request:
Доработал хранение истории согласно ТЗ-6

AntonyukV516 committed:
Добавил новую схему хранения истории,
добавил частичное удаление истории после clear в менеджере,
Добавил новые тесты HistoryManager

AntonyukV516 committed:
Добавил новую схему хранения истории,
добавил частичное удаление истории после clear в менеджере,
Добавил новые тесты HistoryManager

AntonyukV516 committed:
Исправлен code stile

AntonyukV516 committed:
Еще исправлен code stile

AntonyukV516 committed:
Еще исправлен code stile

AntonyukV516 committed:
Еще исправлен code stile

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,59 @@
+import java.util.Map;
 
 public class InMemoryHistoryManager implements HistoryManager {
     private List<PreTask> history = new ArrayList<>();
```
Используется только в методе getHistory(). Надо сделать локальной переменной и перенести объявление в этот метод

vlolad reviewed (changes requested):
Привет!

По тестам все ок, а по новому менеджеру истории нужна доработка. Есть несколько небольших комментариев на улучшение, но основная задача - доработать удаление ноды. См. подсказку №3 из ТЗ.
Метод removeNode должен использоваться как в методе интерфейса remove(), так и add() (если задача добавляется в историю повторно или удаляется)

Жду на повторную проверку, хорошего дня!

AntonyukV516 committed:
Доработаны действия с узлами

AntonyukV516 committed:
Доработаны действия с узлами,
доработал ошибки в тестах

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
+    @Override
+    public void remove(int id) {
+        if (nodeMap.get(id) != null) {
+            Node delitedNode = nodeMap.remove(id);
```
nodeMap.remove(id); можно объединить с функционалом метода removeNode(). В приложении в любом случае из мапы удаление происходит одновременно с разлинковкой ноды

vlolad replied:
Проверку ноды на null в целом тоже можно убрать в  removeNode. В методе тогда останется только вызов removeNode

vlolad commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -7,23 +7,23 @@
     private HashMap<Integer, Epic> epics = new HashMap<>();
     private HashMap<Integer, Task> tasks = new HashMap<>();
     private HashMap<Integer, Subtask> subtasks = new HashMap<>();
     private HistoryManager historyManager = new InMemoryHistoryManager();
```
Надо заменить инициализацию на вызов соответствующего метода из класса Managers. Реализация менеджера истории может измениться

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
-        if (history.size() < HISTORY_MAX_SIZE) {
-            history.addLast(preTask);
-            return preTask;
+    public PreTask add(PreTask preTask) {
```
Метод работает некорректно
Посмотри подсказку №2 в ТЗ, рекомендуется вынести функционал добавления новой ноды в отдельный метод linkLast

Он должен:
1. Создавать новую ноду
2. Если first == null, то first = новая нода
Иначе last.next = новая нода
3. В конце в любом случае меняем last = новая нода

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
-            history.removeFirst();
-            history.addLast(preTask);
-            return preTask;
+            last.previous = oldLast;
```
На этом моменте мы в last.previous линкуем в саму себя, это ошибка (т.к. выше oldLast == last)

Посмотри в дебаггере, как это работает

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
-            return preTask;
+    public PreTask add(PreTask preTask) {
+        Node oldLast = last;
+        Node newNode = new Node(preTask);
```
Перед созданием новой ноды надо также удалить и разлинковать предыдущую ноду с таким же id задачи

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
+        }
+    }
+
+    private void removeNode(Node node) {
```
Метод не работает корректно. У него должны выполняться две задачи:
1. Удалять у переданной ноды связи с next и prev
2. next и prev у переданной ноды линковать между собой

Попутно надо проработать случаи, когда node.prev == null (что значит, что это first-нода). Может быть и так, что из истории удалены все значения, тогда first = last = null (но прямо так писать не надо)

Рекомендую переписать с нуля данный метод, т.к. сейчас логика слишком запутанная. Для работы над ошибками - в 68-69 и 72 строках одинаковые операции

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,83 @@
     public List<PreTask> getHistory() {
-        return new ArrayList<>(history);
+        List<PreTask> history = new ArrayList<>();
+        for (Node node : nodeMap.values()) {
```
Уточнил этот момент - надо поменять порядок формирования списка истории начиная с первой ноды first. nodeMap нам требуется только для получения ноды по id задачи
Примерно таким образом:

> List<PreTask> history = new ArrayList<>();
> Node node = first;
> 
> while (node != null) {
> 	history.add(node.values);
> 	node = node.next;
> 		}

Как раз дополнительная проверка работы истории будет с:

vlolad reviewed (changes requested):
Привет!

Пока история всё ещё плохо работает, линкование новых нод происходит некорректно при добавлении и удалении
Оставил комментарии, но совету перечитать ТЗ и часть основных методов в менеджере истории переделать. Ещё лучше - сделать пару тестовых задач в Main и через дебаггер пройти путь сохранения задач в историю

По тестам пока ок, но после переработки истории надо будет проверить покрытие кода

+ Небольшую недоработку в основном менеджере задач нашел, тоже надо поправить

Если будут вопросы - меня всегда можно найти в Пачке. Хорошего дня!

AntonyukV516 committed:
Переделал с 0 add, removeNode, getHistory
тесты не менял, так как они затрагивают новый функционал,
 но их логика проверок не поменялась

AntonyukV516 committed:
Переделал с 0 add, removeNode, getHistory
тесты не менял, так как они затрагивают новый функционал,
 но их логика проверок не поменялась

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,82 @@
+    }
 
     @Override
     public List<PreTask> getHistory() {
```
Тут всё отлично

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,82 @@
-            history.removeFirst();
-            history.addLast(preTask);
-            return preTask;
+            last.next = newNode;
```
Для новой ноды не назначен previous. Можно здесь добавить `newNode.previous = last;` либо вынести в конструктор

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,82 @@
+    }
+
+    @Override
+    public PreTask add(PreTask preTask) {
```
Тут тоже всё отлично

vlolad commented on src/manager/InMemoryHistoryManager.java:
```diff
@@ -1,31 +1,82 @@
+                node.next.previous = node.previous;
+            }
+            node.previous = null;
+            node.next = null;
```
И лучше node.values = null, так сборщик мусора быстрее удалит лишний объект

Но в целом всё ок, этот метод верно сделан

vlolad reviewed (changes requested):
Привет!

Последняя мелкая правка по методу linkLast() и будем приниматься, сейчас в случае удаления ноды некорректно определяется first, потому что параметр previous не назначается для новых нод

Постараюсь принять ASAP как придет правка. Хороших выходных!

AntonyukV516 committed:
Исправил, согласно замечаниям - linkLast и доработал удаление ноды

vlolad reviewed (approved):
Привет!

Всё отлично, принято
